### DISCUSSION TITLE
Mutating too long

### DISCUSSION THREAD

**Discussion by henrikzabel** [TanStack/query#8045] [Q&A] [answered]
I am currently testing what happens when the server is not online (so there is no connection to the server). The mutation is loading not only super long but the status is also pending. How can I abort the request when tanstack query notices that there is no connection after x seconds?

**Comment by TkDodo** [comment]
if there is no network connection, we pause requests until connection has been established. We return an `isPaused:true` flag in that case

**Reply by henrikzabel** [reply]
I am not talking about the network connection. If the device is offline it works fine. I am talking about the case where the server is offline. The mutation does not even get fired after the server is online again. It's just stuck in the state of `pending`

**Reply by TkDodo** [reply]
so you fire a mutation, and it errors because the server is not available. Then the mutation is "done" from a mutation perspective, and it won't re-fire until you call `mutate()` again.

You can implement `retry` for mutations and have them try periodically until they succeed though

**Reply by henrikzabel** [reply]
But there never is an error. It's just pending until I kill the app

**Reply by TkDodo** [reply]
I'd need to see a minimal reproduction please.

**Reply by henrikzabel** [reply]
Sorry that my reply comes so late.

Following is basically all that you need:

```tsx
import { Button, Text, View } from "react-native";
import { useMutation } from "@tanstack/react-query";

const IndexPage = () => {
  const mutate = useMutation({
    // I have to use my local ip adress here. When using a non-existing url it works as intented
    // (isPending is false and status is error)
    // The server is not running but was in the past.
    mutationFn: async () =>
      fetch("https://192.178.26.5:8000/some-path-that-never-existed", {
        method: "PATCH",
      }),
    onError: (e) => {
      console.log(e);
    },
  });

  return (
    <View
      style={{
        height: "100%",
        backgroundColor: "lime",
        justifyContent: "center",
        paddingHorizontal: 50,
      }}
    >
      <Button onPress={() => mutate.mutate()} title="Mutate" />
      <Text>Status: {mutate.status}</Text>
      <Text>Is Pending: {String(mutate.isPending)}</Text>
    </View>
  );
};

export default IndexPage;
```

The problem only comes with urls that have been mutated once. (or at least that is what I think the problem is)

**Reply by TkDodo** [reply]
This isn't a minimal reproduction, it's a code snippet: https://tkdodo.eu/blog/how-can-i#3-provide-a-minimal-reproducible-example

**Reply by henrikzabel** [reply]
https://github.com/HenrikZabel/tastack-query-issues

Here you go

**Reply by TkDodo** [reply, marked as answer]
not sure what `fetch` is doing here but we can clearly see that the `fetch` call hangs, not the mutation itself:

```
mutationFn: async () => {
    console.log('mutation starts')
  const result = await fetch("https://192.178.26.5:8000/some-path-that-never-existed", {
    method: "PATCH",
  })
    console.log('mutation ends')
    return result
},
```

this will log `mutation starts`, but not `mutation ends`. So the problem lies within the `fetch` call itself, not react query.

My browser devtools also show a never ending `pending` request in the devtools

![Screenshot 2024-09-14 at 23 48 17](https://github.com/user-attachments/assets/cfc2c6a3-995b-4e22-a1ec-10d4d661c827)

so, you have to find out why that is, but it has nothing to do with react-query. You would have the same problem if you just make this `fetch` call without react-query.

**Reply by TkDodo** [reply]
ah, the thing finally error'd out for me after 1.3 minutes
![Screenshot 2024-09-14 at 23 49 49](https://github.com/user-attachments/assets/a497ec05-2c93-4b13-889b-e6fdfa0d136e)

after that, the status correctly goes to error. So, everything 🟢 from react-query perspective
